tests/unittests/algo/test_gridsearch.py

```diff
@@ -0,0 +1,139 @@
+    dim = Integer("yolo", "reciprocal", 1, 1000)
+    assert discrete_grid(dim, 4) == [1, 10, 100, 1000]
+    assert discrete_grid(dim, 6) == [1, 4, 16, 63, 251, 1000]
+
+
+def test_unsupported_discrete_prior():
+    """Test unsupported discrete prior message"""
+    dim = Integer("yolo", "norm", 0, 1)
+    with pytest.raises(TypeError) as exc:
+        discrete_grid(dim, 6)
+
+    assert exc.match("Grid Search only supports `loguniform`, `uniform`")
+
+
+def test_real_grid():
+    """Test linear real grid"""
+    dim = Real("yolo", "uniform", -5, 10)
+    assert real_grid(dim, 6) == [-5, -3, -1, 1, 3, 5]
+    assert real_grid(dim, 7) == [-5 + 10 / 6.0 * i for i in range(7)]
+
+
+def test_log_real_grid():
+    """Test logarithmic real grid"""
+    dim = Real("yolo", "reciprocal", 0.0001, 1)
+    assert numpy.allclose(real_grid(dim, 5), [0.0001, 0.001, 0.01, 0.1, 1])
+    assert real_grid(dim, 6) == [
+        numpy.exp(numpy.log(0.0001) + (numpy.log(1) - numpy.log(0.0001)) / 5 * i)
+        for i in range(6)
+    ]
+
+
+def test_unsupported_real_prior():
+    """Test unsupported real prior message"""
+    dim = Real("yolo", "norm", 0.0001, 1)
+    with pytest.raises(TypeError) as exc:
+        real_grid(dim, 5)
+
+    assert exc.match("Grid Search only supports `loguniform`, `uniform`")
+
+
+def test_build_grid():
+    """Test that grid search builds the proper grid"""
+    dim1 = Real("dim1", "uniform", 0, 1)
+    dim2 = Integer("dim2", "uniform", 0, 10)
+    dim3 = Categorical("dim3", "abcde")
+    space = Space()
+    space.register(dim1)
+    space.register(dim2)
+    space.register(dim3)
+
+    grid = GridSearch.build_grid(space, {"dim1": 3, "dim2": 4, "dim3": 1})
+    assert len(grid) == 3 * 4 * 5
+
+
+def test_build_grid_limit_size(caplog):
+    """Test that grid search reduces the n_values when grid is too large"""
+    dim1 = Real("dim1", "uniform", 0, 1)
+    dim2 = Integer("dim2", "uniform", 0, 10)
+    dim3 = Categorical("dim3", "abcde")
+    space = Space()
+    space.register(dim1)
+    space.register(dim2)
+    space.register(dim3)
+
+    with caplog.at_level(logging.WARNING, logger="orion.algo.gridsearch"):
+        grid = GridSearch.build_grid(space, {k: 5 for k in space}, 100)
+    assert len(grid) == 4 * 4 * 5
+
+    assert "`n_values` reduced by 1 to limit number of trials below 100" in caplog.text
+
+
+def test_build_grid_cannot_limit_size(caplog):
+    """Test that when choices are too large GridSearch raises ValueError"""
+    dim1 = Real("dim1", "uniform", 0, 1)
+    dim2 = Integer("dim2", "uniform", 0, 10)
+    dim3 = Categorical("dim3", "abcde")
+    dim4 = Categorical("dim4", "abcde")
+    space = Space()
+    space.register(dim1)
+    space.register(dim2)
+    space.register(dim3)
+    space.register(dim4)
+
+    with pytest.raises(ValueError) as exc:
+        GridSearch.build_grid(space, {k: 5 for k in space}, 10)
+
+    assert exc.match(
+        "Cannot build a grid smaller than 10. "
```

Comment:
Same as above, for the space, there could be a grid with size `8`, where first 2 dimension with only one candidate. Is it valid?

Reply:
The source of the (intentional) problem is the 2 categorical dimensions with 5 choices each. The grid always picks all choices, no matter what `n_values` is. This leads to a smallest possible grid of 1 * 1 * 5 * 5 > 10.

Reply:
oh, I made a mistake, thought `Categorical("dim3", "abcde")` had 2 values, but it had 5.